# Parsing Retrosheet Event Files

There are more questions I have about trends in homerun statistics that unfortunately can't be answered by the Baseball Databank datasets. Instead, I want to turn to the far more detailed Retrosheet datasets, so that I can answer more specific questions. For example, people are always talking about Coors Field in Denver as being a scary place for pitchers--something about the high altitude and/or shape of the stadium as making it especially homerun-prone. Is it true?

To answer this kind of question, the simple offensive statlines provided by the Baseball Databank datasets simply aren't enough. What we really need are the Retrosheet event files, which preserve historical baseball games inning by inning and, in some cases, pitch by pitch.

The problem is that they're not formatted in a way that can immediately be used but need to be parsed. (<a href="https://www.retrosheet.org/eventfile.htm">Here's</a> some info on how these event files are structured.)

So, as a preliminary step, I need to build some parsing functions that can help us deal with Retrosheet's huge amounts of information.

Let me show you what I mean.

I started by downloading event file datasets for 2010–2021—here's a glimpse:

In [1]:
import os

os.listdir('../data/retrosheet/reg_season/')[:20]

['2021CHA.EVA',
 '2010CIN.EVN',
 '2014SEA.EVA',
 'TBA2016.ROS',
 'SLN2015.ROS',
 'LAN2015.ROS',
 'CIN2011.ROS',
 'CHN2011.ROS',
 '2014CLE.EVA',
 'TOR2021.ROS',
 'ARI2020.ROS',
 '2019MIN.EVA',
 '2019TEX.EVA',
 '2014SFN.EVN',
 'SEA2017.ROS',
 '2015NYA.EVA',
 '2015TEX.EVA',
 '2015MIN.EVA',
 'BOS2011.ROS',
 'OAK2010.ROS']

There's two types of datasets here:
<ul>
    <li> Event files are prefixed by the year and then the home team to which they pertain, appended by EVA or EVN depending on whether it's an American League or National League team. With 30 teams, there will be 30 of these per year/season.</li>
    <li> Roster files are appended by .ROS, and these just contain roster info for the entire season—not relevant for us now. For each year, there will be one .ROS file per team.</li>
</ul>

Here's a glimpse of what one of these event files looks like:

In [2]:
path = '../data/retrosheet/reg_season/2017LAN.EVN'

with open(path) as f:
    print(f.read()[:1300])

id,LAN201704030
version,2
info,visteam,SDN
info,hometeam,LAN
info,site,LOS03
info,date,2017/04/03
info,number,0
info,starttime,1:10PM
info,daynight,day
info,usedh,false
info,umphome,vanol901
info,ump1b,marqa901
info,ump2b,fairc901
info,ump3b,rackd901
info,howscored,park
info,pitches,pitches
info,oscorer,whitj701
info,temp,67
info,winddir,torf
info,windspeed,3
info,fieldcond,unknown
info,precip,unknown
info,sky,sunny
info,timeofgame,172
info,attendance,53701
info,wp,kersc001
info,lp,chacj001
info,save,
start,margm001,"Manuel Margot",0,1,8
start,myerw001,"Wil Myers",0,2,3
start,solay001,"Yangervis Solarte",0,3,4
start,renfh001,"Hunter Renfroe",0,4,9
start,schir001,"Ryan Schimpf",0,5,5
start,hedga001,"Austin Hedges",0,6,2
start,aybae001,"Erick Aybar",0,7,6
start,chacj001,"Jhoulys Chacin",0,8,1
start,jankt001,"Travis Jankowski",0,9,7
start,tolea001,"Andrew Toles",1,1,7
start,seagc001,"Corey Seager",1,2,6
start,turnj001,"Justin Turner",1,3,5
start,gonza003,"Adrian Gonzalez",1,4,3
start,fors

There are different kinds of records here, identified by the term that prepending each row.
<ul>
    <li>The <code>id</code> record at the top is a unique identifer for each game</li>
    <li>The <code>info</code> records pertain to game-specific data, such as the home team, visiting team, attendance, etc.</li>
    <li>The <code>start</code> records specify the players who are starting for both teams</li>
    <li>the <code>play</code> records specify individual plays</li>
</ul>
And then there are a number of other record types that we don't see in this snippet here.

For a simple place to start, I decided to write some routines that will parse the <code>id</code> and <code>info</code> record types. This won't allow us to answer my question about Coors Field, but it's a first step in figuring out how to make these datasets useable.

In the <code>retrosheet_utils</code> module I'll import below, I've written two functions thus far:
<ul>
    <li><code>parse_event_file_info()</code> takes a path to a sepcific event file and returns a list of dictionaries, each of which contains <code>info</code> fields for a given game <code>id</code></li>
    <li><code>load_season_info()</code> takes a year and runs <code>parse_event_file_info()</code> for each event file pertaining to that year, returning a list of dictionaries of all the games of that year</li>
    </ul>
 
Here's what it looks like:           

In [3]:
import retrosheet_utils

data = retrosheet_utils.load_season_info(2017)

data[:5]

[{'id': 'BOS201704030',
  'visteam': 'PIT',
  'hometeam': 'BOS',
  'site': 'BOS07',
  'date': '2017/04/03',
  'number': '0',
  'starttime': '2:06PM',
  'daynight': 'day',
  'usedh': 'true',
  'umphome': 'demud901',
  'ump1b': 'nauep901',
  'ump2b': 'guccc901',
  'ump3b': 'torrc901',
  'howscored': 'park',
  'pitches': 'pitches',
  'oscorer': 'shalm701',
  'temp': '48',
  'winddir': 'fromrf',
  'windspeed': '13',
  'fieldcond': 'unknown',
  'precip': 'unknown',
  'sky': 'cloudy',
  'timeofgame': '183',
  'attendance': '36594',
  'wp': 'porcr001',
  'lp': 'coleg001',
  'save': 'kimbc001'},
 {'id': 'BOS201704050',
  'visteam': 'PIT',
  'hometeam': 'BOS',
  'site': 'BOS07',
  'date': '2017/04/05',
  'number': '0',
  'starttime': '7:10PM',
  'daynight': 'night',
  'usedh': 'true',
  'umphome': 'nauep901',
  'ump1b': 'guccc901',
  'ump2b': 'torrc901',
  'ump3b': 'demud901',
  'howscored': 'park',
  'pitches': 'pitches',
  'oscorer': 'shalm701',
  'temp': '40',
  'winddir': 'fromrf',
  'winds

With 30 teams and 162 regular season games (each of which, of course, involves 2 teams!), the above list should contain info for 30 * 162 / 2 = 2430 games.

In [4]:
len(data)

2400

Seems we're missing 30 games, but we'll figure that out later.

In this format, in any case, we can easily create a DataFrame.

In [5]:
import pandas as pd
import numpy as np

In [6]:
df = pd.DataFrame.from_records(data)
df

id visteam hometeam   site        date number starttime  \
0     BOS201704030     PIT      BOS  BOS07  2017/04/03      0    2:06PM   
1     BOS201704050     PIT      BOS  BOS07  2017/04/05      0    7:10PM   
2     BOS201704110     BAL      BOS  BOS07  2017/04/11      0    7:10PM   
3     BOS201704120     BAL      BOS  BOS07  2017/04/12      0    7:10PM   
4     BOS201704130     PIT      BOS  BOS07  2017/04/13      0    2:06PM   
...            ...     ...      ...    ...         ...    ...       ...   
2395  CHN201709150     SLN      CHN  CHI11  2017/09/15      0    1:21PM   
2396  CHN201709160     SLN      CHN  CHI11  2017/09/16      0    3:05PM   
2397  CHN201709170     SLN      CHN  CHI11  2017/09/17      0    1:23PM   
2398  CHN201709290     CIN      CHN  CHI11  2017/09/29      0    1:22PM   
2399  CHN201709300     CIN      CHN  CHI11  2017/09/30      0    3:05PM   

     daynight  usedh   umphome  ...  winddir windspeed fieldcond   precip  \
0         day   true  demud901  ...   fromrf        13   unknown  unknown   
1       night   true  nauep901  ...   fromrf         8   unknown  unknown   
2       night   true  coope901  ...     tolf        16   unknown  unknown   
3       night   true  johna901  ...     rtol         6   unknown  unknown   
4         day   true  morag901  ...     torf        15   unknown  unknown   
...       ...    ...       ...  ...      ...       ...       ...      ...   
2395      day  false  bakej902  ...     rtol        11   unknown  unknown   
2396      day  false  torrc901  ...     rtol        11   unknown  unknown   
2397      day  false  drecb901  ...   fromrf         5   unknown  unknown   
2398      day  false  muchm901  ...  unknown        15   unknown  unknown   
2399      day  false  wintm901  ...   fromrf        11   unknown  unknown   

          sky timeofgame attendance        wp        lp      save  
0      cloudy        183      36594  porcr001  coleg001  kimbc001  
1     unknown        233      36137  kellj001  basta001      None  
2     unknown        195      37497  pomed001  bundd001      None  
3     unknown        226      32211  givem001  wrigs001      None  
4      cloudy        195      32400  barnm001  nicaj001  kimbc001  
...       ...        ...        ...       ...       ...       ...  
2395    sunny        187      38464  edwac001  martc006      None  
2396    sunny        169      40959  hendk001  wachm001  daviw001  
2397  unknown        224      37242  strop001  lyont001  daviw001  
2398    sunny        176      36258  duenb001  lorem002  grimj002  
2399    sunny        184      41493  lestj001  stepj002      None  

[2400 rows x 27 columns]

In this format, we can actually make some sense of all this information!

For example, in 2017, which ballparks had the highest average attendance?

In [7]:
df['attendance'] = df['attendance'].astype('int')

avg_attendance = df.groupby('site')['attendance'].mean().sort_values(ascending=False)

avg_attendance

site
LOS03    46482.287500
STL10    42544.375000
SFO03    40810.062500
CHI11    39482.387500
TOR02    39456.150000
NYC21    39383.600000
ANA01    37308.062500
DEN02    36508.800000
BOS07    36039.512500
MIL06    31294.120482
WAS11    31116.600000
ARL02    30922.487500
ATL03    30900.862500
HOU03    30570.415584
NYC20    29897.771084
DET05    28721.462500
KAN06    27351.162500
PHO01    26429.087500
SEA03    26396.875000
SAN02    26376.900000
MIN04    25289.137500
CLE08    25226.275000
BAL12    25062.500000
PIT08    23950.278481
PHI13    23495.000000
CIN09    22642.150000
MIA02    20231.064935
CHI12    20123.425000
OAK01    18282.362500
STP01    15013.550000
WIL02     2596.000000
Name: attendance, dtype: float64

We can make this a little more legible by joining with a table containing the ballpark names that match the codes"

In [8]:
ballparks = pd.read_csv('../data/retrosheet/parkcodes.csv')
ballparks

PARKID                                 NAME  \
0    ALB01                       Riverside Park   
1    ALT01                        Columbia Park   
2    ANA01             Angel Stadium of Anaheim   
3    ARL01                    Arlington Stadium   
4    ARL02        Rangers Ballpark in Arlington   
..     ...                                  ...   
255  WIL02        BB&T Ballpark at Bowman Field   
256  WNY01     West New York Field Club Grounds   
257  WOR01   Agricultural County Fair Grounds I   
258  WOR02  Agricultural County Fair Grounds II   
259  WOR03       Worcester Driving Park Grounds   

                                          AKA           CITY STATE  \
0                                         NaN         Albany    NY   
1                                         NaN        Altoona    PA   
2               Edison Field; Anaheim Stadium        Anaheim    CA   
3                                         NaN      Arlington    TX   
4    The Ballpark in Arlington; Ameriquest Fl      Arlington    TX   
..                                        ...            ...   ...   
255                                       NaN   Williamsport    PA   
256                                       NaN  West New York    NJ   
257                                       NaN      Worcester    MA   
258                                       NaN      Worcester    MA   
259                                       NaN      Worcester    MA   

          START         END LEAGUE  \
0    09/11/1880  05/30/1882     NL   
1    04/30/1884  05/31/1884     UA   
2    04/19/1966         NaN     AL   
3    04/21/1972  10/03/1993     AL   
4    04/11/1994  09/29/2019     AL   
..          ...         ...    ...   
255  08/20/2017  08/20/2017     NL   
256  09/11/1898  09/17/1899     NL   
257  05/01/1880  09/29/1882     NL   
258  08/17/1887  08/17/1887     NL   
259  10/30/1874  10/30/1874    NaN   

                                                 NOTES  
0       TRN:9/11/80;6/15&9/10/1881;5/16-5/18&5/30/1882  
1                                                  NaN  
2                                                  NaN  
3                                                  NaN  
4                                                  NaN  
..                                                 ...  
255                                                PIT  
256  BRO:9/18&10/2/1898; NY1:9/11/98, 6/4&7/16&8/13...  
257                                                NaN  
258                                         1 BSN game  
259                                         1 BS1 game  

[260 rows x 9 columns]

In [9]:
pd.merge(left=avg_attendance, right=ballparks, how='left', left_on='site', right_on='PARKID')

attendance PARKID                                       NAME  \
0   46482.287500  LOS03                             Dodger Stadium   
1   42544.375000  STL10                          Busch Stadium III   
2   40810.062500  SFO03                                  AT&T Park   
3   39482.387500  CHI11                              Wrigley Field   
4   39456.150000  TOR02                              Rogers Centre   
5   39383.600000  NYC21                          Yankee Stadium II   
6   37308.062500  ANA01                   Angel Stadium of Anaheim   
7   36508.800000  DEN02                                Coors Field   
8   36039.512500  BOS07                                Fenway Park   
9   31294.120482  MIL06                                Miller Park   
10  31116.600000  WAS11                             Nationals Park   
11  30922.487500  ARL02              Rangers Ballpark in Arlington   
12  30900.862500  ATL03                              Suntrust Park   
13  30570.415584  HOU03                           Minute Maid Park   
14  29897.771084  NYC20                                 Citi Field   
15  28721.462500  DET05                              Comerica Park   
16  27351.162500  KAN06                           Kauffman Stadium   
17  26429.087500  PHO01                                Chase Field   
18  26396.875000  SEA03                               Safeco Field   
19  26376.900000  SAN02                                 PETCO Park   
20  25289.137500  MIN04                               Target Field   
21  25226.275000  CLE08                          Progressive Field   
22  25062.500000  BAL12                Oriole Park at Camden Yards   
23  23950.278481  PIT08                                   PNC Park   
24  23495.000000  PHI13                         Citizens Bank Park   
25  22642.150000  CIN09                    Great American Ballpark   
26  20231.064935  MIA02                               Marlins Park   
27  20123.425000  CHI12  Guaranteed Rate Field;U.S. Cellular Field   
28  18282.362500  OAK01            Oakland-Alameda County Coliseum   
29  15013.550000  STP01                            Tropicana Field   
30   2596.000000  WIL02              BB&T Ballpark at Bowman Field   

                                         AKA            CITY STATE  \
0                              Chavez Ravine     Los Angeles    CA   
1                                        NaN       St. Louis    MO   
2                Pacific Bell Park; SBC Park   San Francisco    CA   
3                   Weeghman Park; Cubs Park         Chicago    IL   
4                                    Skydome         Toronto   ONT   
5                                        NaN        New York    NY   
6              Edison Field; Anaheim Stadium         Anaheim    CA   
7                                        NaN          Denver    CO   
8                                        NaN          Boston    MA   
9                                        NaN       Milwaukee    WI   
10                                       NaN      Washington    DC   
11  The Ballpark in Arlington; Ameriquest Fl       Arlington    TX   
12                                       NaN         Atlanta    GA   
13                 Enron Field; Astros Field         Houston    TX   
14                                       NaN        New York    NY   
15                                       NaN         Detroit    MI   
16                            Royals Stadium     Kansas City    MO   
17                         Bank One Ballpark         Phoenix    AZ   
18                                       NaN         Seattle    WA   
19                                       NaN       San Diego    CA   
20                                       NaN     Minneapolis    MN   
21                              Jacobs Field       Cleveland    OH   
22                                       NaN       Baltimore    MD   
23                                       NaN      Pittsburgh    PA   
24             

There you have it! In 2017, Dodger Stadium (go Dodgers!) had the highest average attendence at 46,482, followed by the Cardinals' Busch Stadium, the Giants' AT&T Park, and the Cubs' Wrigley Field.

Maybe we'll also want to have a look at things like weather and wind conditions, game duration, etc. On their own these might be fun to look at, but once the rest of the Retrosheet event file is parsed, we could have some potentially powerful insights on our hands about how game conditions correlate with game stats.